### ISSUE TITLE
expressions in parameter fail to evaluate

### ISSUE THREAD

**Issue by dkriegner** [xraypy/xraylarch#185]
possibly I hit a bug while trying to create parameters with mathematical expressions which use a mathematical function (like *sqrt*, *sin*) without further basic mathematic operation inside the function. 

The following minimal example illustrates the problem:
```
p = group()
_sys.paramGroup=p

p.x = param(0.5, vary=False)
p.z = param(expr='sin(1*x)')
p.z2 = param(expr='sin(x)')
show(p)
```

The output of the program in the *larch* command line is the following:
```
$ larch
==============================================================================
Larch 0.9.33 (2017-Apr-3) M. Newville, M. Koker, B. Ravel, and others
Python 2.7.13 (default, Feb 20 2017, 10:54:54) 
[GCC 5.4.0],
numpy 1.12.1, scipy 0.18.1, matplotlib 1.5.3, wxpython 3.0.2.0
==============================================================================

larch> p = group()
larch> _sys.paramGroup=p
larch> 
larch> p.x = param(0.5, vary=False)
larch> p.z = param(expr='sin(1*x)')
larch> p.z2 = param(expr='sin(x)')
larch> show(p)
== Group 0x7f2c9a60ac50: 0 methods, 3 attributes ==
  x: param(value=0.50000, vary=False)
  z: param(value=0.47943, vary=False, expr='sin(1*x)')
  z2: param(value=None, vary=False, expr='sin(x)')
Traceback (most recent calls last): 
    sin(x)
   File <stdin>, line 1
  File "/usr/lib64/python2.7/site-packages/larch/interpreter.py", line 810, in on_call
    return func(*args, **keywords)

AttributeError: 'Parameter' object has no attribute 'sin'
larch>
```

While the expression _sin(1*x)_ evaluates as expected _sin(x)_ causes an error. Is it wrong to assume that both should work?

This was tested on the current larch git master (commit 1d2b3b8)

**Comment by newville** [comment]
@dkriegner Thanks -- yes, I can reproduce this.  I'm not 100% sure of the cause, but the solution is already underway. In fact, the issue is very timely, as PR #184 reworks all the underlying fitting code to use lmfit.   This is close to ready.

With #184, the above code does work, but the expressions will be stored but not "live" updated (that is your p.y will not have a value of 0.479...).    The expressions would work when fitting, but this issue brings up good point of missing functionality in the new version.  

So, there will be a new type/subclass of called `param_group()` added  that *will* have live-updated values, and also work with fitting.

FWIW, the `_sys.paramGroup` business will go away -- the "namespace" for expressions during fits will have lots and lots of functionality and will include all the parameters, but won't include everything in Larch's namespace.

**Comment by dkriegner** [comment]
@newville So currently I should wait for #184 to be merged? 

To give you the bigger picture of what I try to accomplish: I am trying to use some hacks around the old mkfit script and extract the generated equations (for atomic distances) to be used as expressions in a EXAFS fit using larch. Of course it would be also nice if one could evaluate the expressions/generate the corresponding EXAFS path separate from a fit. 

Do you think I could try to achieve that with what is in #184 currently?

**Comment by newville** [comment]
@dkriegner  I think waiting for #184 is probably the fastest option for a fix.  Currently (like, as of 10 minutes ago), I get on that branch:

```
p = param_group()
p.x = param(0.5, vary=True)
p.y1 = param(expr='sin(x)')
p.y2 = param(expr='sin(1.0*x)')

show(p)

== Param Group 0x1145a8750: 0 methods, 4 attributes ==
  _params: Parameters([('x', <Parameter 'x', 0.5, bounds=[-inf:inf]>), ('y1', <Parameter 'y1', 0.47942553860420301, bounds=[-inf:inf], expr='sin(x)'>), ('y2', <Parameter 'y2', 0.47942553860420301, bounds=[-inf:inf], expr='sin(1.0*x)'>)])
  x: <Parameter 'x', 0.5, bounds=[-inf:inf]>
  y1: <Parameter 'y1', 0.47942553860420301, bounds=[-inf:inf], expr='sin(x)'>
  y2: <Parameter 'y2', 0.47942553860420301, bounds=[-inf:inf], expr='sin(1.0*x)'>

```

Where `_params` is the `lmfit.Parameters()`, encapsulating the individual `Parameter` objects, and managing their relationship (so that `y` knows what `x` is).  

It would be interesting to try to migrate or somehow incorporate `mkfit` into Larch.  I haven't looked at that in several years, and don't know that I have a copy of it.  Maybe @bruceravel knows or has an opinion on this?

**Comment by dkriegner** [comment]
@newville I will wait for your fixes and meanwhile define everything as `sin(0+variable)` 

my current approach is clearly a dirty hack: I run the original `mkfit ` script and later parse the resulting `equations.inp` file to generate a long list of parameters in larch. A clean implementation of this would certainly be better.

**Comment by newville** [comment]
I believe this is now fixed in master, and soon to be released version 0.9.34 (hopefully within a few days).
